server/kernel_cmdline.go

```diff
@@ -0,0 +1,130 @@
+package server
+
+// See
+// https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/ctype.h
+// and
+// https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/lib/ctype.c
+func isspace(ch byte) bool {
+	// Row 8 - 15
+	if ch >= 9 && ch <= 13 {
+		return true
+	} else if ch == 32 {
+		return true
+	} else if ch == 160 {
+		return true
+	}
+	return false
+}
+
+func skipSpaces(str []byte) []byte {
+	i := 0
+	for i < len(str) && isspace(str[i]) {
+		i++
+	}
+	return str[i:]
+}
+
+// This function should replicate the exact behavior of `next_arg` in
+// https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/lib/cmdline.c
+// nextArg takes in a command line and returns the next space-separated argument
+// in the line. It ignores spaces enclosed in quotes.
+// nextArg returns the param and value, if the argument has an '='. It then returns
+// the remaining command line to run nextArg on.
+//
+// Note:
+// - nextArg *mutates* args. Make a copy prior to using this function.
+// - args *must* be null-terminated (e.g., the final byte is 0x00).
+// - args cannot be the C-style empty string (e.g., {0x00}) or empty byte slice.
+// - Consider using parseArgs instead as it does not have these limitations.
+func nextArg(args []byte) (param []byte, val []byte, remaining []byte) {
+	var i uint
+	var equals uint
+	inQuote := false
+	quoted := false
+
+	if args[0] == '"' {
+		args = args[1:]
+		inQuote = true
+		quoted = true
+	}
+
+	for i = 0; len(args) > int(i) && args[i] != 0; i++ {
+		if isspace(args[i]) && !inQuote {
+			break
+		}
+		if equals == 0 {
+			if args[i] == '=' {
+				equals = i
+			}
+		}
+		if args[i] == '"' {
+			inQuote = !inQuote
+		}
+	}
+
+	param = args[:i]
+	if equals == 0 {
+		val = nil
+	} else {
+		args[equals] = 0
+		param = param[:equals]
+		val = args[equals+1 : i]
+
+		/* Don't include quotes in value. */
+		if val[0] == '"' {
+			val = val[1:]
+			if args[i-1] == '"' {
+				args[i-1] = 0
+				val = val[:len(val)-1]
+			}
+		}
+	}
+
+	if quoted && i > 0 && args[i-1] == '"' {
+		args[i-1] = 0
+		// Since we don't know whether args[i-1] is in param or val, we
+		// use equals to differentiate.
+		if equals == 0 {
+			param = param[:len(param)-1]
+		} else {
+			val = val[:len(val)-1]
+		}
+	}
+
+	if args[i] != 0 {
+		args[i] = 0
```

Comment:
I saw this in the kernel C code as well, but I don't quite understand why need to assign 0 to args[i]

Reply:
Idk whether you're asking why the kernel does it or why we do it, so I'll explain both:
The kernel sets the null terminator at the point i because that byte represents the next space found in args. As a result, the val/param return values in the kernel can be properly null-terminated but still point to the same memory as in args initially (no more mem allocs). 

For golang: we try and modify args as closely as the kernel does. Even if golang doesn't have C-style/null-terminated strings, we interpret the byte array the same way, since we also have logic that deals with if any byte in arg == 0.

Reply:
I see arg[i] should be a space before, so it can be set to 0.